# How To Drop Constant Features Using Variance Threshold

Video Link: https://www.youtube.com/watch?v=uMlU2JaiOd8&list=PLZoTAELRMXVPgjwJ8VyRoqmfNs2CJwhVH&index=1

In [78]:
import pandas as pd

In [79]:
data=pd.DataFrame({
    'A': [1,2,4,1,3],
    'B': [4,5,6,7,8],
    'C': [0,0,0,0,0],
    'D': [1,1,1,1,1]
})

In [80]:
data.head()

A  B  C  D
0  1  4  0  1
1  2  5  0  1
2  4  6  0  1
3  1  7  0  1
4  3  8  0  1

**Observations:**

Here we can observe values in C and D have Constants features. So we can remove this 2 columns, for that purpose we use a sklearn package called as **Variance Threshold**.

More on: 
https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.VarianceThreshold.html



### Variance Threshold
is a Feature selector that removes all low-variance features. This feature selection algorithm looks only at the features (X), not the desired outputs (y), and can thus be used for unsupervised learning.

**Syntax:**

skearn.feature_selection.VarianceThreshold(threshold=0.0)

If we gave threshold value as 0, then it will go and check each and every feature which has variance as 0. Ie. whichever columns has variance 0, it will be removed.

Here column C has same values and column D has same values. If we compute mean, variance or standard deviation the value will be same as thier column value, because all the values are same.




In [81]:
from sklearn.feature_selection import VarianceThreshold

In [82]:
threshold = VarianceThreshold()
threshold.fit(data)

VarianceThreshold()

In [83]:
threshold.get_support()

array([ True,  True, False, False])

**Inference:**

Here True represents feature A and B, while False represents feature C and D. Next we pick up the columns C and D that having False value. Because those are the columns needs to be removed. C and D column has 0 variance. They are not important for training.

In [84]:
data.columns[threshold.get_support()] 

Index(['A', 'B'], dtype='object')

<b>Get the columns that need to be removed<b>

In [85]:
constant_columns = [column for column in data.columns if column not in data.columns[threshold.get_support()]]
constant_columns

['C', 'D']

<b>Drop the constant columns<b>

In [86]:
data.drop(constant_columns, axis=1)

A  B
0  1  4
1  2  5
2  4  6
3  1  7
4  3  8

## Practice on Bigger Dataset

Dataset Source: https://www.kaggle.com/competitions/santander-customer-satisfaction/data?select=train.csv



### 1. Load the dataset

In [87]:
df = pd.read_csv('santander.csv', nrows=10000)
df.head()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   
2   4     2     23                 0.0                      0.0   
3   8     2     37                 0.0                    195.0   
4  10     2     39                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  ...  \
0                        0                        0  ...   
1                        0                        0  ...   
2                        0                        0  ...   
3                        0                        0  ...   
4                        0                        0  ...   

   saldo_medio_var33_hace2  saldo_medio_var33_hace3  saldo_medio_var33_ult1  \
0                      0.0                      0.0                     0.0   
1                      0.0                      0.0                     0.0   
2                      0.0                      0.0                     0.0   
3                      0.0                      0.0                     0.0   
4                      0.0                      0.0                     0.0   

   saldo_medio_var33_ult3  saldo_medio_var44_hace2  saldo_medio_var44_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   
2                     0.0                      0.0                      0.0   
3                     0.0                      0.0                      0.0   
4                     0.0                      0.0                      0.0   

   saldo_medio_var44_ult1  saldo_medio_var44_ult3          var38  TARGET  
0                     0.0                     0.0   39205.170000       0  
1                     0.0                     0.0   49278.030000       0  
2                     0.0                     0.0   67333.770000       0  
3                     0.0                     0.0   64007.970000       0  
4                     0.0                     0.0  117310.979016       0  

[5 rows x 371 columns]

In [88]:
df.shape

(10000, 371)

### 2. Dependent and Independent features

In [89]:
X=df.drop(['TARGET'], axis=1) # independent features
X.head()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   
2   4     2     23                 0.0                      0.0   
3   8     2     37                 0.0                    195.0   
4  10     2     39                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  ...  \
0                        0                        0  ...   
1                        0                        0  ...   
2                        0                        0  ...   
3                        0                        0  ...   
4                        0                        0  ...   

   saldo_medio_var29_ult3  saldo_medio_var33_hace2  saldo_medio_var33_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   
2                     0.0                      0.0                      0.0   
3                     0.0                      0.0                      0.0   
4                     0.0                      0.0                      0.0   

   saldo_medio_var33_ult1  saldo_medio_var33_ult3  saldo_medio_var44_hace2  \
0                     0.0                     0.0                      0.0   
1                     0.0                     0.0                      0.0   
2                     0.0                     0.0                      0.0   
3                     0.0                     0.0                      0.0   
4                     0.0                     0.0                      0.0   

   saldo_medio_var44_hace3  saldo_medio_var44_ult1  saldo_medio_var44_ult3  \
0                      0.0                     0.0                     0.0   
1                      0.0                     0.0                     0.0   
2                      0.0                     0.0                     0.0   
3                      0.0                     0.0                     0.0   
4                      0.0                     0.0                     0.0   

           var38  
0   39205.170000  
1   49278.030000  
2   67333.770000  
3   64007.970000  
4  117310.979016  

[5 rows x 370 columns]

In [90]:
y=df[['TARGET']] # dependent feature
y.head()

TARGET
0       0
1       0
2       0
3       0
4       0

### 3. Split the dataset into training and test set.

In [91]:
from sklearn.model_selection import train_test_split

In [92]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.30,random_state=0)

In [93]:
X_train.head()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
7681  15431     2     42               840.0                  4477.02   
9031  18181     2     31                 0.0                    52.32   
3691   7411     2     51                 0.0                     0.00   
202     407     2     36                 0.0                     0.00   
5625  11280     2     23                 0.0                     0.00   

      imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  \
7681                  4989.54                      0.0   
9031                    52.32                      0.0   
3691                     0.00                      0.0   
202                      0.00                      0.0   
5625                     0.00                      0.0   

      imp_op_var40_comer_ult3  imp_op_var40_efect_ult1  \
7681                      0.0                        0   
9031                      0.0                        0   
3691                      0.0                        0   
202                       0.0                        0   
5625                      0.0                        0   

      imp_op_var40_efect_ult3  ...  saldo_medio_var29_ult3  \
7681                        0  ...                     0.0   
9031                        0  ...                     0.0   
3691                        0  ...                     0.0   
202                         0  ...                     0.0   
5625                        0  ...                     0.0   

      saldo_medio_var33_hace2  saldo_medio_var33_hace3  \
7681                      0.0                      0.0   
9031                      0.0                      0.0   
3691                      0.0                      0.0   
202                       0.0                      0.0   
5625                      0.0                      0.0   

      saldo_medio_var33_ult1  saldo_medio_var33_ult3  saldo_medio_var44_hace2  \
7681                     0.0                     0.0                      0.0   
9031                     0.0                     0.0                      0.0   
3691                     0.0                     0.0                      0.0   
202                      0.0                     0.0                      0.0   
5625                     0.0                     0.0                      0.0   

      saldo_medio_var44_hace3  saldo_medio_var44_ult1  saldo_medio_var44_ult3  \
7681                      0.0                     0.0                     0.0   
9031                      0.0                     0.0                     0.0   
3691                      0.0                     0.0                     0.0   
202                       0.0                     0.0                     0.0   
5625                      0.0                     0.0                     0.0   

          var38  
7681   37491.21  
9031  106685.94  
3691   66144.66  
202    92121.36  
5625   74650.83  

[5 rows x 370 columns]

In [94]:
X_train.shape

(7000, 370)

In [95]:
y_test.head()

TARGET
9394       0
898        0
2398       0
5906       0
2343       0

### 4. Apply Variance threshold feature selector

In [96]:
var_thresh = VarianceThreshold()
var_thresh.fit(X_train)

VarianceThreshold()

### 5. Find non constant feature

In [97]:
var_thresh.get_support()

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True, False, False,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True, False, False,  True,  True,  True,  True,  True, False,
       False,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True, False, False, False, False,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
       False, False,  True,  True,  True,  True,  True,  True,  True,
       False,  True,  True,  True, False, False,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True, False, False,  True,  True,  True,
        True,  True, False, False,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,

In [98]:
sum(var_thresh.get_support())

284

**Inference**

Total of 284 feature are not constant. So out of 370 features, 284 are not constant and remaining 86 are constant features.

### 6. Get the constant columns

In [99]:
constant_features = [column for column in X_train.columns if column not in X_train.columns[var_thresh.get_support()]]
# [column for column in data.columns if column not in data.columns[threshold.get_support()]]
constant_features

['ind_var2_0',
 'ind_var2',
 'ind_var13_medio_0',
 'ind_var13_medio',
 'ind_var18_0',
 'ind_var18',
 'ind_var27_0',
 'ind_var28_0',
 'ind_var28',
 'ind_var27',
 'ind_var34_0',
 'ind_var34',
 'ind_var41',
 'ind_var46_0',
 'ind_var46',
 'num_var13_medio_0',
 'num_var13_medio',
 'num_var18_0',
 'num_var18',
 'num_var27_0',
 'num_var28_0',
 'num_var28',
 'num_var27',
 'num_var34_0',
 'num_var34',
 'num_var41',
 'num_var46_0',
 'num_var46',
 'saldo_var13_medio',
 'saldo_var18',
 'saldo_var28',
 'saldo_var27',
 'saldo_var34',
 'saldo_var41',
 'saldo_var46',
 'delta_imp_amort_var18_1y3',
 'delta_imp_amort_var34_1y3',
 'delta_imp_reemb_var17_1y3',
 'delta_imp_reemb_var33_1y3',
 'delta_imp_trasp_var17_out_1y3',
 'delta_imp_trasp_var33_out_1y3',
 'delta_num_reemb_var17_1y3',
 'delta_num_reemb_var33_1y3',
 'delta_num_trasp_var17_out_1y3',
 'delta_num_trasp_var33_out_1y3',
 'imp_amort_var18_hace3',
 'imp_amort_var18_ult1',
 'imp_amort_var34_hace3',
 'imp_amort_var34_ult1',
 'imp_var7_emit_ult1',
 

**Inference**

These are the columns having zero variance.

### 7. Drop the constant features from training set

In [100]:
X_train.drop(constant_features, axis=1, inplace=True)

In [101]:
X_train.shape

(7000, 284)

**Inference**

After all the filtering, we have 284 features in training set

In [102]:
X_train.head()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
7681  15431     2     42               840.0                  4477.02   
9031  18181     2     31                 0.0                    52.32   
3691   7411     2     51                 0.0                     0.00   
202     407     2     36                 0.0                     0.00   
5625  11280     2     23                 0.0                     0.00   

      imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  \
7681                  4989.54                      0.0   
9031                    52.32                      0.0   
3691                     0.00                      0.0   
202                      0.00                      0.0   
5625                     0.00                      0.0   

      imp_op_var40_comer_ult3  imp_op_var40_efect_ult1  \
7681                      0.0                        0   
9031                      0.0                        0   
3691                      0.0                        0   
202                       0.0                        0   
5625                      0.0                        0   

      imp_op_var40_efect_ult3  ...  saldo_medio_var29_ult3  \
7681                        0  ...                     0.0   
9031                        0  ...                     0.0   
3691                        0  ...                     0.0   
202                         0  ...                     0.0   
5625                        0  ...                     0.0   

      saldo_medio_var33_hace2  saldo_medio_var33_hace3  \
7681                      0.0                      0.0   
9031                      0.0                      0.0   
3691                      0.0                      0.0   
202                       0.0                      0.0   
5625                      0.0                      0.0   

      saldo_medio_var33_ult1  saldo_medio_var33_ult3  saldo_medio_var44_hace2  \
7681                     0.0                     0.0                      0.0   
9031                     0.0                     0.0                      0.0   
3691                     0.0                     0.0                      0.0   
202                      0.0                     0.0                      0.0   
5625                     0.0                     0.0                      0.0   

      saldo_medio_var44_hace3  saldo_medio_var44_ult1  saldo_medio_var44_ult3  \
7681                      0.0                     0.0                     0.0   
9031                      0.0                     0.0                     0.0   
3691                      0.0                     0.0                     0.0   
202                       0.0                     0.0                     0.0   
5625                      0.0                     0.0                     0.0   

          var38  
7681   37491.21  
9031  106685.94  
3691   66144.66  
202    92121.36  
5625   74650.83  

[5 rows x 284 columns]